independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -151,60 +147,11 @@ public ProtectionDomain getProtectionDomain() {
         return protectionDomain;
     }
 
-    private JarFile readLockAcquireAndGetJarReference() {
-        while (true) {
-            readLock.lock();
-            final JarFile zipFileLocal = this.zipFile;
-            if (zipFileLocal != null) {
-                //Expected fast path: returns a reference to the open JarFile while owning the readLock
-                return zipFileLocal;
-            } else {
-                //This Lock implementation doesn't allow upgrading a readLock to a writeLock, so release it
-                //as we're going to need the WriteLock.
-                readLock.unlock();
-                //trigger the JarFile being (re)opened.
-                ensureJarFileIsOpen();
-                //Now since we no longer own any lock, we need to try again to obtain the readLock
-                //and check for the reference still being valid.
-                //This exposes us to a race with closing the just-opened JarFile;
-                //however this should be extremely rare, so we can trust we won't loop much;
-                //A counter doesn't seem necessary, as in fact we know that methods close()
-                //and resetInternalCaches() are invoked each at most once, which limits the amount
-                //of loops here in practice.
-            }
-        }
-    }
-
-    private void ensureJarFileIsOpen() {
-        writeLock.lock();
-        try {
-            if (this.zipFile == null) {
-                try {
-                    this.zipFile = JarFiles.create(jarPath.toFile());
-                } catch (IOException e) {
-                    throw new RuntimeException("Failed to open " + jarPath, e);
-                }
-            }
-        } finally {
-            writeLock.unlock();
-        }
-    }
-
     @Override
     public void close() {
-        writeLock.lock();
-        try {
-            final JarFile zipFileLocal = this.zipFile;
-            if (zipFileLocal != null) {
-                try {
-                    this.zipFile = null;
-                    zipFileLocal.close();
-                } catch (Throwable e) {
-                    //ignore
-                }
-            }
-        } finally {
-            writeLock.unlock();
+        var ref = jarFileReference.get();
+        if (ref != null) {
+            ref.join().close(this);
```

Comment:
in theory join shouldn't wait here: you could use ref.getNow() and adding an assert that you always expect some value here, because the file should be already populated due to the ref count logic (which should close only if there are no acquisition pending)